# 07 - SQL Analysis
## Credit Risk Analysis — Give Me Some Credit

**Objetivo de este notebook:**
- Crear una base de datos MySQL con el dataset limpio
- Cargar los datos en la base de datos
- Realizar consultas SQL para analizar el riesgo crediticio
- Extraer insights de negocio mediante SQL

In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine, text
import pymysql

# Cargamos el dataset con features
df = pd.read_csv('../data/processed/cs-training-features.csv')
print(f"Dataset cargado: {df.shape}")

Dataset cargado: (149731, 12)


In [2]:
from dotenv import load_dotenv
import os

# Cargamos el .env desde la raíz del proyecto
load_dotenv(dotenv_path='../.env')

password = os.getenv('MYSQL_PASSWORD')
print(f"Password cargada: {'Sí' if password else 'No'}")

Password cargada: Sí


## Paso 1- Conexión a MySQL y creación de la base de datos

In [3]:
engine = create_engine(f'mysql+pymysql://root:{password}@localhost:3306')

with engine.connect() as conn:
    conn.execute(text("CREATE DATABASE IF NOT EXISTS credit_risk"))
    conn.commit()
    print("Base de datos creada")

engine = create_engine(f'mysql+pymysql://root:{password}@localhost:3306/credit_risk')
print("Conexión establecida")

Base de datos creada
Conexión establecida


## Paso 2 — Cargar los datos en MySQL

In [4]:
# Cargamos el dataset en MySQL
df.to_sql('credit_risk_data', con=engine, if_exists='replace', index=False)
print(f"Tabla creada con {len(df)} registros")

Tabla creada con 149731 registros


## Paso 3 — Consultas SQL de análisis

In [5]:
# Consulta 1: Cúal es la distribución de la variable objetivo
query = """
SELECT seriousdlqin2yrs, COUNT(*) as total
FROM credit_risk_data
GROUP BY seriousdlqin2yrs
"""
resultado = pd.read_sql(query, con=engine)
print(resultado)

   seriousdlqin2yrs   total
0                 1    9879
1                 0  139852


### Consulta 1 — Distribución de la variable objetivo
- Los 139.852 clientes no tuvieron problemas de pago
- 9.879 clientes sí tuvieron problemas de pago
- Confirma el desbalanceo de clases que tratamos con SMOTE en el modelado

In [6]:
# Consulta 2: Cual es la edad media de los clientes con y sin problemas de pago
query = """
SELECT seriousdlqin2yrs, AVG(age) as edad_media
FROM credit_risk_data
GROUP BY seriousdlqin2yrs
"""
resultado = pd.read_sql(query, con=engine)
print(resultado)

   seriousdlqin2yrs  edad_media
0                 1     46.0515
1                 0     52.7600


### Consulta 2 — Edad media de los clientes con y sin problemas de pago
- La edad media de los clientes que no tienen problemas de pago es de 52 años
- La edad media de los clientes que tienen problemas de pago es de 46 años
- Confirma que los clientes problemáticos son más jovenes como vimos en el EDA

In [7]:
# Consulta 3: Cuál es el ingreso mensual medio de los clientes con y sin problemas de pago
query = """
SELECT seriousdlqin2yrs, AVG(monthlyincome) as ingreso_medio
FROM credit_risk_data
GROUP BY seriousdlqin2yrs
"""
resultado = pd.read_sql(query, con=engine)
print(resultado)

   seriousdlqin2yrs  ingreso_medio
0                 1    5257.265513
1                 0    5923.905118


### Consulta 3 — Ingresos mensuales medios de clientes con y sin problemas de pago
- El ingreso medio de los clientes que no tienen problemas de pago es de 5.924€
- El ingreso medio de los clientes que tienen problemas de pago es de 5.257€
- Confirma que los clientes problemáticos tienen menos ingresos mensuales de media, como vimos en el EDA

In [8]:
# Consulta 4: Cuántos clientes tienen alto riesgo crediticio
query = """
SELECT credit_utilization_risk, COUNT(*) as total_clientes
FROM credit_risk_data
GROUP BY credit_utilization_risk
"""
resultado = pd.read_sql(query, con=engine)
print(resultado)

   credit_utilization_risk  total_clientes
0                        1           62005
1                        0           87726


### Consulta 4 — Clientes que tienen alto riesgo crediticio
- Hay 62.005 clientes con riesgo alto crediticio
- Hay 87.726 clientes sin riesgo alto crediticio

In [9]:
# Consulta 5: Promedio de retrasos totales para clientes con y sin problemas de pago
query = """
SELECT seriousdlqin2yrs, AVG(total_delays) as media_retrasos
FROM credit_risk_data
GROUP BY seriousdlqin2yrs
"""
resultado = pd.read_sql(query, con=engine)
print(resultado)

   seriousdlqin2yrs  media_retrasos
0                 1          2.0295
1                 0          0.2860


### Consulta 5 — Promedio de retrasos totales para clientes con y sin problemas de pago
- Los clientes con problemas de pago tienen una media de 2.03 retrasos
- Los clientes sin problemas de pago tienen una media de 0.29 retrasos
- sto confirma que total_delays es una variable muy importante para predecir el impago, como vimos en Feature Importance

In [11]:
# Consulta 6: Cuales son lo s10 clientes con mayor debtratio que tienen problemas de pago
query = """
SELECT seriousdlqin2yrs, debtratio, age, monthlyincome, total_delays
FROM credit_risk_data
WHERE seriousdlqin2yrs = 1
ORDER BY debtratio DESC
LIMIT 10
"""
resultado = pd.read_sql(query, con=engine)
print(resultado)

   seriousdlqin2yrs  debtratio  age  monthlyincome  total_delays
0                 1       1.91   58         5400.0             4
1                 1       1.91   48         5400.0             5
2                 1       1.91   38         5400.0             2
3                 1       1.91   44         5400.0             0
4                 1       1.91   38            1.0             0
5                 1       1.91   73         5400.0             5
6                 1       1.91   47         5400.0             2
7                 1       1.91   35         5400.0             0
8                 1       1.91   58         5400.0             2
9                 1       1.91   63         5400.0             0


### Consulta 6 — Top 10 clientes con mayor debtratio y problemas de pago
- Todos tienen debtratio = 1.91 ,el límite del capping aplicado en la limpieza
- La mayoría tiene monthlyincome tienen 5400€, que es la mediana usada para imputar nulos
- Sus edades varían entre 35 y 73 años
- Tienen entre 0 y 5 retrasos totales
- Estos perfiles representan clientes de alto riesgo por sobreendeudamiento
- La fila 4 tiene monthlyincome 1.0,uno de los valores bajos que decidimos mantener en la limpieza por no tener una referencia razonable para imputar.